### PR TITLE
[BEAM-4132] Default type hints for multi-output transforms

### PR DESCRIPTION
**Pull request by udim**
For tagged multi-output results, the contained PCollections' `element_type` members are left as `None`. This fix recursively runs type inference on these PCollections.
See bug for examples.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
R: @robertwb, @aaltay

File: sdks/python/apache_beam/pipeline.py [hunk 1]
```diff
@@ -571,6 +573,10 @@ def _infer_result_type(self, transform, inputs, result_pcollection):
       else:
         result_pcollection.element_type = transform.infer_output_type(
             input_element_type)
+    elif isinstance(result_pcollection, pvalue.DoOutputsTuple):
+      # Single-input, multi-output inference.
+      for pcoll in result_pcollection:
+        self._infer_result_type(transform, inputs, pcoll)
```
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: How does this infer a different type to the different collections?

File: sdks/python/apache_beam/typehints/typed_pipeline_test.py [hunk 2]
```diff
@@ -131,6 +131,41 @@ def filter_fn(data):
...
+        if element % 2:
+          yield beam.pvalue.TaggedOutput('odd', element)
+        else:
+          yield beam.pvalue.TaggedOutput('even', element)
+
+    p = TestPipeline()
+    res = (p
+           | beam.Create([1, 2, 3])
+           | beam.ParDo(MyDoFn()).with_outputs('odd', 'even'))
+    # Test that the element type of even and odd is int.
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typed_pipeline_test.py]: Sure it's not Any?

File: sdks/python/apache_beam/pipeline.py (hunk 1 above)
Comment (aaltay) [reply sdks/python/apache_beam/pipeline.py]: I do not believe it does, I missed this while reviewing.

I have a more fundamental question. Is it possible to infer types for multiple outputs without explicit typehints from the user? DoFn will choose their taggedoutput dynamically based on the element. And I am not sure how we can infer different output types only by knowing the input type and all list of possible return types from a function. In other words, is it possible to infer output types (without explicit hints) to be anything other than union of all possible output types from that transform?

File: sdks/python/apache_beam/typehints/typed_pipeline_test.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/typed_pipeline_test.py]: More precisely it should read that we're testing it's not None and consistent with int.

File: sdks/python/apache_beam/pipeline.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/pipeline.py]: This indeed does set the same output type for each pcoll. Maybe there's a way to infer more deeply. I can look at this later if this would be of value (I'm thinking that we should come up with a way to specify multi-output type hints).
Currently this is better that leaving pcoll.element_type as None, which fails type checking.

Comment (udim) [reply sdks/python/apache_beam/pipeline.py]: Looks like the best bet right is to hard-code multi-outputs to Any.

Comment (robertwb) [reply sdks/python/apache_beam/pipeline.py]: That would fix the bug that they're currently None, but the TODO would still need to be left in place.

Comment (udim) [reply sdks/python/apache_beam/pipeline.py]: Done; thanks for the input!
